File: backend/src/pipeline.py
"""End-to-end orchestrator: company name -> CBE number -> filings -> financials."""

from __future__ import annotations

import logging
from dataclasses import dataclass
from typing import Callable, Optional

from src._timing import timed
from src.config import Settings, settings as default_settings
from src.extraction import build_extractor
from src.kbo import KBOScraper
from src.models import Company, CompanyFinancialReport
from src.nbb import NBBClient

logger = logging.getLogger(__name__)

ProgressCallback = Callable[[str], None]


@dataclass(slots=True)
class PipelineConfig:
    filings_to_read: int = 2


class EnrichmentPipeline:
    def __init__(
        self,
        *,
        settings: Settings | None = None,
        config: PipelineConfig | None = None,
    ) -> None:
        self._settings = settings or default_settings
        self._config = config or PipelineConfig()

    def run(
        self,
        query: str,
        on_progress: Optional[ProgressCallback] = None,
        *,
        filings_to_read: Optional[int] = None,
    ) -> CompanyFinancialReport:
        """Resolve *query* into a :class:`CompanyFinancialReport`.

        Args:
            query: Name or 10-digit CBE.
            on_progress: Optional per-step progress hook.
            filings_to_read: Override for ``PipelineConfig.filings_to_read``
                so the API can pass a per-request value without rebuilding
                the singleton pipeline.
        """
        with timed("pipeline.total"):
            return self._run(query, on_progress, filings_to_read)

    def _run(
        self,
        query: str,
        on_progress: Optional[ProgressCallback],
        filings_to_read: Optional[int],
    ) -> CompanyFinancialReport:
        notify = on_progress or (lambda _msg: None)
        n_filings = (
            filings_to_read
            if filings_to_read is not None
            else self._config.filings_to_read
        )

        notify("Resolving company in KBO")
        company = self._resolve_company(query)
        logger.info(
            "Resolved %r -> %s (%s)",
            query,
            company.name or "<name unknown, KBO skipped>",
            company.enterprise_number,
        )

        with NBBClient(
            base_url=self._settings.nbb_api_base_url,
            subscription_key=self._settings.nbb_api_subscription_key,
            cache_dir=self._settings.cache_dir,
            timeout=self._settings.http_timeout,
            deposit_path=self._settings.nbb_deposit_path,
        ) as nbb:
            notify("Fetching filing references from NBB")
            references = nbb.latest_references(
                company.enterprise_number, limit=n_filings
            )
            logger.info(
                "Fetched %d filing reference(s) for %s",
                len(references),
                company.enterprise_number,
            )

            extractor = build_extractor(nbb)
            chain_label = "XBRL → " + (
                f"LLM ({self._settings.openai_model}) → regex"
                if self._settings.has_openai_credentials
                else "regex"
            )
            notify(f"Extractor chain: {chain_label}")
            logger.info("Using chain: %s", chain_label)

            statements = []
            for i, ref in enumerate(references, 1):
                notify(f"Extracting filing {i}/{len(references)} (ref {ref.reference})")
                stmt = extractor.extract(company.enterprise_number, ref)
                if stmt is not None:
                    statements.append(stmt)

        notify("Done")
        return CompanyFinancialReport(
            company=company, filings=references, statements=statements
        )

    def _resolve_company(self, query: str) -> Company:
        """Return a :class:`Company` for *query*.

        ``KBOScraper.lookup`` handles both forms: when *query* is already
        a valid 10-digit enterprise number it skips the name search and
        fetches the KBO detail page directly (one HTTP call). When *query*
        is a name it does the phonetic search first, then the detail page.

        Either way the returned :class:`Company` carries the legal profile
        (NACE codes, functions, characteristics) — number-based input no
        longer short-circuits past KBO. If KBO is unreachable, the call
        raises ``CompanyNotFoundError`` and the pipeline aborts.
        """
        with KBOScraper(timeout=self._settings.http_timeout) as kbo, timed("kbo.lookup_total"):
            return kbo.lookup(query)

File: frontend/lib/server/pipeline.ts
/**
 * End-to-end orchestrator: company name -> CBE -> filings -> financials.
 *
 * Active extraction path: PDF + OpenAI structured outputs (see
 * extraction/pdf.ts). For every NBB filing reference we download the
 * PDF deliverable, run pdf-parse to get the text, and ask OpenAI to
 * fill the FinancialStatement schema. The XBRL extractor is kept on
 * disk (extraction/xbrl.ts) but no longer wired into the pipeline.
 */

import { KBOScraper } from "./kbo/scraper";
import { NBBClient } from "./nbb/client";
import { NoFilingsError } from "./errors";
import { PdfExtractor } from "./extraction/pdf";
import { createLogger } from "./log";
import {
  type Company,
  type CompanyFinancialReport,
  type FinancialStatement,
} from "./models";

const log = createLogger("pipeline");

export type ProgressCallback = (msg: string) => void;

export type PipelineOptions = {
  filingsToRead?: number;
  onProgress?: ProgressCallback;
};

const DEFAULT_FILINGS_TO_READ = 3;

export class EnrichmentPipeline {
  /**
   * Resolve *query* (name or 10-digit CBE) into a full report.
   *
   * KBO ambiguity surfaces as {@link AmbiguousMatchError} from the
   * scraper — callers convert it to a 409. KBO not-found surfaces as
   * a generic KBOScraperError; the route handler maps it to 404.
   */
  async run(
    query: string,
    opts?: PipelineOptions,
  ): Promise<CompanyFinancialReport> {
    const notify = opts?.onProgress ?? (() => {});
    const nFilings = opts?.filingsToRead ?? DEFAULT_FILINGS_TO_READ;

    log.info("run start", { query, filings: nFilings });
    const startTotal = performance.now();

    notify("Resolving company in KBO");
    const kbo = new KBOScraper();
    const company: Company = await log.time("kbo.lookup", () => kbo.lookup(query));
    log.info("kbo resolved", {
      cbe: company.enterprise_number,
      name: company.name,
      nace: company.nace_codes.length,
      functions: company.functions.length,
      corporate_mandates: company.corporate_mandates.length,
    });

    notify("Fetching filing references from NBB");
    const nbb = new NBBClient();
    const references = await log.time("nbb.latestReferences", () =>
      nbb.latestReferences(company.enterprise_number, nFilings),
    );
    log.info("nbb references", {
      cbe: company.enterprise_number,
      count: references.length,
      refs: references.map((r) => r.reference),
    });

    if (references.length === 0) {
      log.warn("no filings", { cbe: company.enterprise_number });
      throw new NoFilingsError({
        enterprise_number: company.enterprise_number,
        name: company.name,
      });
    }

    notify(`Extractor: PDF+LLM (${references.length} filing(s))`);
    const extractor = new PdfExtractor(nbb);

    const statements: FinancialStatement[] = [];
    let i = 0;
    for (const ref of references) {
      i++;
      notify(`Extracting filing ${i}/${references.length} (ref ${ref.reference})`);
      const stmt = await log.time(`extract ${i}/${references.length} ref=${ref.reference}`, () =>
        extractor.extract(company.enterprise_number, ref),
      );
      if (stmt !== null) {
        statements.push(stmt);
      } else {
        log.warn("extract returned null", { ref: ref.reference });
      }
    }

    notify("Done");
    log.info("run end", {
      cbe: company.enterprise_number,
      filings: references.length,
      statements: statements.length,
      ms: Math.round(performance.now() - startTotal),
    });
    return { company, filings: references, statements };
  }
}
